### PR TITLE
Re-implement handling of `RolloverCompleted` for taker

### PR DESCRIPTION
**Pull request by thomaseizinger**

### CODE REVIEW

File: daemon/src/collab_settlement_maker.rs [hunk 1]
```diff
@@ -18,7 +17,7 @@ use xtra_productivity::xtra_productivity;
 pub struct Actor {
     cfd: Cfd,
     projection: xtra::Address<projection::Actor>,
-    on_completed: Box<dyn MessageChannel<Completed<CollaborativeSettlement>>>,
+    on_completed: Box<dyn MessageChannel<CollaborativeSettlementCompleted>>,
```
Comment (da-kami) [inline daemon/src/collab_settlement_maker.rs:21]: Note: I think the creation of the alias is somehow not part of this commit although it should be. This just shows the usage of the alias but not the alias itself.

Sidenote: 
We had this alias before, but I thought we agreed that we don't need to introduce one for collaborative settlement 😅

File: daemon/src/model/cfd.rs [hunk 2]
```diff
@@ -1414,16 +1415,16 @@ pub mod marker {
...
+pub type SetupCompleted = Completed<(Dlc, marker::Setup), anyhow::Error>;
 
 /// Message sent from a rollover actor to the
 /// cfd actor to notify that the rollover has finished (contract got updated).
 /// TODO: Roll it out in the maker rollover actor
-pub type RolloverCompleted = Completed<(Dlc, marker::Rollover)>;
+pub type RolloverCompleted = Completed<(Dlc, marker::Rollover), anyhow::Error>;
 
-pub type CollaborativeSettlementCompleted = Completed<CollaborativeSettlement>;
+pub type CollaborativeSettlementCompleted = Completed<CollaborativeSettlement, anyhow::Error>;
```
Comment (da-kami) [inline daemon/src/model/cfd.rs:1504]: ah, here it is :D

File: daemon/src/rollover_taker.rs [hunk 3]
```diff
@@ -305,6 +308,30 @@ impl Actor {
         .await;
     }
 
+    pub async fn handle_rollover_timeout_reached(
```
Comment (da-kami) [inline daemon/src/rollover_taker.rs:304]: 👍

File: daemon/src/auto_rollover.rs [hunk 4]
```diff
@@ -112,8 +115,43 @@ where
...
+
+        let dlc = match event.event {
+            CfdEvent::RolloverCompleted { dlc } => dlc,
+            CfdEvent::RolloverFailed | CfdEvent::RolloverRejected => {
+                return Ok(());
+            }
+            _ => bail!("Unexpected event {:?}", event.event),
+        };
+
+        tracing::info!("Setup complete, publishing on chain now");
```
Comment (da-kami) [inline daemon/src/auto_rollover.rs]:
```suggestion
        tracing::info!("Rollover completed");
```

This message is a copy-paste from contract-setup and incorrect in this context.

File: daemon/src/collab_settlement_maker.rs (hunk 1 above)
Comment (thomaseizinger) [reply daemon/src/collab_settlement_maker.rs:21]: I reintroduced it because adding the error type everywhere was annoying :D

Comment (thomaseizinger) [reply daemon/src/collab_settlement_maker.rs:21]: Thanks for pointing out the messy commit, will fix!

File: daemon/src/auto_rollover.rs (hunk 4 above)
Comment (thomaseizinger) [reply daemon/src/auto_rollover.rs]: Thanks for pointing it out!

File: daemon/src/rollover_taker.rs [hunk 5]
```diff
@@ -176,6 +177,11 @@ impl Actor {
 
         ctx.stop();
     }
+
+    /// Returns whether the maker has accepted our rollover proposal.
+    fn is_accepted(&self) -> bool {
+        self.rollover_msg_sender.is_some()
```
Comment (luckysori) [inline daemon/src/rollover_taker.rs:184]: Hehe, nice heuristic.

File: daemon/src/auto_rollover.rs [hunk 6]
```diff
@@ -139,10 +139,7 @@ where
 {
     async fn started(&mut self, ctx: &mut xtra::Context<Self>) {
         let fut = ctx
-            .notify_interval(
-                rollover_taker::MAX_ROLLOVER_DURATION + Duration::from_secs(60),
-                || AutoRollover,
-            )
+            .notify_interval(Duration::from_secs(5 * 60), || AutoRollover)
```
Comment (luckysori) [inline daemon/src/auto_rollover.rs:157]: ❓ Why the magic number `5`?

File: daemon/src/rollover_taker.rs [hunk 7]
```diff
@@ -160,18 +159,18 @@ impl Actor {
...
 
-    async fn update_proposal(
-        &self,
-        proposal: Option<(RolloverProposal, SettlementKind)>,
-    ) -> Result<()> {
-        self.projection
+    /// Update the proposal state in the projection actor.
+    ///
+    /// This function is non-fallible because we don't care if the actor goes away, it is not
+    /// critical to continue with the protocol.
```
Comment (luckysori) [inline daemon/src/rollover_taker.rs:165]: 👍 Good learning for the rest of the ephemeral actors.

File: daemon/src/rollover_taker.rs [hunk 8]
```diff
@@ -137,7 +137,8 @@ impl Actor {
 
         let this = ctx.address().expect("self to be alive");
         self.tasks.add(async move {
-            // Use an explicit type annotation to cause a compile error if someone changes the handler.
+            // Use an explicit type annotation to cause a compile error if someone changes the
+            // handler.
```
Comment (luckysori) [inline daemon/src/rollover_taker.rs:141]: 🙃 Mega-nit, but this belongs in the previous commit.

**Review by luckysori — APPROVED** [review]
Very easy to review with all the small commits. LGTM!

File: daemon/src/auto_rollover.rs (hunk 6 above)
Comment (thomaseizinger) [reply daemon/src/auto_rollover.rs:157]: This just restores the previous functionality. `MAX_ROLLOVER_DURATION` was 4 minutes, I just made this not depend on this constant anymore because the internals now work differently.

File: daemon/src/rollover_taker.rs (hunk 7 above)
Comment (thomaseizinger) [reply daemon/src/rollover_taker.rs:165]: Well, the projection actor should likely be gone from these actors completely!

Comment (luckysori) [reply daemon/src/rollover_taker.rs:165]: That is the dream!

**Comment by thomaseizinger** [comment]
I am currently testing this locally. Will report back with the results. It should definitely be a step forward so feel free to merge @luckysori @klochowicz !

**Comment by thomaseizinger** [comment]
> I am currently testing this locally. Will report back with the results. It should definitely be a step forward so feel free to merge @luckysori @klochowicz !

Not quite working as expected but at least it doesn't crash anything!

```
2021-12-30 18:11:50 DEBUG daemon::model::cfd: Rollover was not started: The Cfd was just rolled over order_id=7a00c9cf-9896-4836-b782-c5d1eb4da3c1
2021-12-30 18:16:50 DEBUG daemon::model::cfd: Rollover was not started: The Cfd was just rolled over order_id=7a00c9cf-9896-4836-b782-c5d1eb4da3c1
2021-12-30 18:21:50 DEBUG daemon::model::cfd: Rollover was not started: The Cfd was just rolled over order_id=7a00c9cf-9896-4836-b782-c5d1eb4da3c1
2021-12-30 18:26:50 DEBUG daemon::model::cfd: Rollover was not started: The Cfd was just rolled over order_id=7a00c9cf-9896-4836-b782-c5d1eb4da3c1
2021-12-30 18:28:30 DEBUG daemon::monitor: Got notification for new block block_height=2133660
2021-12-30 18:31:50 DEBUG daemon::model::cfd: Rollover was not started: The Cfd was just rolled over order_id=7a00c9cf-9896-4836-b782-c5d1eb4da3c1
2021-12-30 18:36:50 DEBUG daemon::model::cfd: Rollover was not started: The Cfd was just rolled over order_id=7a00c9cf-9896-4836-b782-c5d1eb4da3c1
2021-12-30 18:41:50 DEBUG daemon::model::cfd: Rollover was not started: The Cfd was just rolled over order_id=7a00c9cf-9896-4836-b782-c5d1eb4da3c1
2021-12-30 18:46:50 DEBUG daemon::model::cfd: Rollover was not started: The Cfd was just rolled over order_id=7a00c9cf-9896-4836-b782-c5d1eb4da3c1
2021-12-30 18:48:30 DEBUG daemon::monitor: Got notification for new block block_height=2133661
2021-12-30 18:51:50 DEBUG daemon::model::cfd: Rollover was not started: The Cfd was just rolled over order_id=7a00c9cf-9896-4836-b782-c5d1eb4da3c1
2021-12-30 18:56:50 DEBUG daemon::model::cfd: Rollover was not started: The Cfd was just rolled over order_id=7a00c9cf-9896-4836-b782-c5d1eb4da3c1
2021-12-30 19:00:00 DEBUG daemon::oracle: Fetching announcement for /x/BitMEX/BXBT/2021-12-31T09:00:00.price?n=20
2021-12-30 19:01:50 DEBUG daemon::model::cfd: Rollover was not started: The Cfd was just rolled over order_id=7a00c9cf-9896-4836-b782-c5d1eb4da3c1
2021-12-30 19:06:50 DEBUG daemon::model::cfd: Rollover was not started: The Cfd was just rolled over order_id=7a00c9cf-9896-4836-b782-c5d1eb4da3c1
2021-12-30 19:08:50 DEBUG daemon::monitor: Got notification for new block block_height=2133662
2021-12-30 19:11:50 DEBUG daemon::model::cfd: Rollover was not started: The Cfd was just rolled over order_id=7a00c9cf-9896-4836-b782-c5d1eb4da3c1
2021-12-30 19:16:50 DEBUG daemon::model::cfd: Rollover was not started: The Cfd was just rolled over order_id=7a00c9cf-9896-4836-b782-c5d1eb4da3c1
2021-12-30 19:21:50 DEBUG daemon::model::cfd: Rollover was not started: The Cfd was just rolled over order_id=7a00c9cf-9896-4836-b782-c5d1eb4da3c1
2021-12-30 19:26:50 DEBUG daemon::model::cfd: Rollover was not started: The Cfd was just rolled over order_id=7a00c9cf-9896-4836-b782-c5d1eb4da3c1
2021-12-30 19:28:50 DEBUG daemon::monitor: Got notification for new block block_height=2133663
2021-12-30 19:31:50 DEBUG daemon::model::cfd: Rollover was not started: The Cfd was just rolled over order_id=7a00c9cf-9896-4836-b782-c5d1eb4da3c1
```

**Comment by klochowicz** [comment]
@thomaseizinger @luckysori perhaps we could close this PR if we've cherry-picked all the relevant bits?

**Comment by luckysori** [comment]
Woops, forgot to close it. Thanks, @klochowicz!
